frontend/ui/widget/htmlboxwidget.lua

```diff
@@ -14,15 +14,200 @@ local logger = require("logger")
+    end
+
+    local found_start = false
+    local words = {}
+    local rects = {}
+
+    for line_index = start_line_index, end_line_index do
+        local line = lines[line_index]
+        local line_last_rect = nil
+        local line_text_direction = getLineTextDirection(line)
+
+        for word_index, word in ipairs(line) do
+            if type(word) == 'table' then
+                if line_index == start_line_index and word_index == start_word_index then
+                    found_start = true
+                end
+
+                if found_start then
+                    table.insert(words, word.word)
+
+                    -- do not try to join word rects in mixed direction lines
+                    if line_last_rect == nil or line_text_direction == 0 then
+                        local rect = Geom:new{
+                            x = word.x0,
+                            y = line.y0,
+                            w = word.x1 - word.x0,
+                            h = line.y1 - line.y0,
+                        }
+
+                        table.insert(rects, rect)
+                        line_last_rect = rect
+                    else
+                        if line_text_direction > 0 then -- left to right
+                            line_last_rect.w = word.x1 - line_last_rect.x
+                        else -- right to left
+                            line_last_rect.w = line_last_rect.w + (line_last_rect.x - word.x0)
+                            line_last_rect.x = word.x0
+                        end
+                    end
+
+                    if line_index == end_line_index and word_index == end_word_index then
+                        break
+                    end
+                end
+            end
+        end
+    end
+
+    if found_start then
+        return table.concat(words, " "), rects
+    else
+        return nil, nil
+    end
+end
+
+local function areTextBoxesEqual(boxes1, text1, boxes2, text2)
+    if text1 ~= text2 then
+        return false
+    end
+
+    if boxes1 and boxes2 then
+        if #boxes1 ~= #boxes2 then
+            return false
+        end
+
+        for i = 1, #boxes1, 1 do
+            if boxes1[i] ~= boxes2[i] then
+                return false
+            end
+        end
+
+        return true
+    else
+        return (boxes1 == nil) == (boxes2 == nil)
+    end
+end
+
 local HtmlBoxWidget = InputContainer:extend{
     bb = nil,
     dimen = nil,
+    dialog = nil, -- parent dialog that will be set dirty
     document = nil,
     page_count = 0,
     page_number = 1,
+    page_boxes = nil,
     hold_start_pos = nil,
+    hold_end_pos = nil,
     hold_start_time = nil,
     html_link_tapped_callback = nil,
+
+    -- If highlight_text_selection is true then text selection will be highlighted.
+    -- If false then text selection will work as it did originally: no highlighting, and the
+    -- text selection will be updated only on hold release (so hold time might be incorrect).
```

Comment:
Is `(so hold time might be incorrect)` still valid ? (Thought you had fixed something related while we discussed it.)

`and the text selection will be updated only on hold release` reads odd: on hold release, there is nothing different than when `highlight_text_selection = true`, we just don't highlight it, and in both cases, we update text_selection as we pan - and when we release, we return the last update of text_selection, isn't it?

Both here and in textboxwidget.lua.

Reply:
Nice catch. I forgot to update the comments.